packages/nightingale-track-canvas/src/utils/utils.ts

```diff
@@ -0,0 +1,208 @@
+        const element = array[i];
+        if (predicate(element, i, array)) return element;
+    }
+    return undefined;
+}
+
+
+/** Return index of the first element of `sortedArray` for which `key(element) >= query`.
+ * Return length of `sortedArray` if `key(element) < query` for all elements.
+ * (aka Return the first index where `query` could be inserted while keeping the array sorted.) */
+export function firstGteqIndex<T>(sortedArray: ArrayLike<T>, query: number, key: (element: T) => number) {
+    return firstGteqIndexInRange(sortedArray, query, 0, sortedArray.length, key);
+}
+
+/** Return index of the first element within range [start, end) for which `key(element) >= query`.
+ * Return `end` if for all elements in the range `key(element) < query`. */
+function firstGteqIndexInRange<T>(sortedArray: ArrayLike<T>, query: number, start: number, end: number, key: (element: T) => number) {
+    if (start === end) return start;
+    // Invariants:
+    // key(sortedArray[i]) < query for each i < start
+    // key(sortedArray[i]) >= query for each i >= end
+    while (end - start > 4) {
+        const mid = (start + end) >> 1;
+        if (key(sortedArray[mid]) >= query) {
+            end = mid;
+        } else {
+            start = mid + 1;
+        }
+    }
+    // Linear search remaining 4 or fewer elements:
+    for (let i = start; i < end; i++) {
+        if (key(sortedArray[i]) >= query) {
+            return i;
+        }
+    }
+    return end;
+}
+
+
+/** Data structure for storing integer intervals (ranges) and efficiently retrieving a subset of ranges which overlap with another (query) interval.
+ * Not suited for storing float intervals, but query interval can be float. */
+export class RangeCollection<T> {
+    protected readonly items: T[];
+    protected readonly starts: number[];
+    protected readonly stops: number[];
+
+    /** Keys to `this.bins`, sorted in ascending order */
+    protected readonly binSpans: number[];
+    /** `this.bins[span]` contains indices of all items whose length is `<= span` but `> span/Q`, in ascending order */
+    protected readonly bins: Record<number, number[]>;
+    /** Ratio of spans of neighboring bins */
+    protected readonly Q = 2;
+    /** Reusable arrays (one for each bin), to avoid repeated array allocation */
+    private readonly _tmpArrays: Record<number, number[]> = {};
+
+    /** Create a new collection of ranges. `start` must return range start (inclusive), `stop` must return range end (exclusive) */
+    constructor(items: T[], accessors: { start: (item: T) => number, stop: (item: T) => number }) {
+        const { start, stop } = accessors;
+        this.items = items;
+        this.starts = items.map(start);
+        this.stops = items.map(stop);
+
+        this.bins = {};
+        for (let i = 0; i < items.length; i++) {
+            const length = this.stops[i] - this.starts[i];
+            let binSpans = 1;
+            while (binSpans < length) binSpans *= this.Q;
```

Comment:
If you wanted to avoid the while loop you could do
```
const binSpans = this.Q ** Math.ceil(Math.log(length) / Math.log(this.Q));
```
and even pull out the final log above the for loop eg:
```
const logQ = Math.log(this.Q);
...
const binSpans = this.Q ** Math.ceil(Math.log(length) / logQ);
```

Reply:
I agree it would be faster, but I wanted to avoid float calculations because of imprecisions.
For some numbers, `Math.ceil(Math.log(length) / Math.log(this.Q))` could give one more or one less than expected. Trying to fix that would include some ifs or whiles anyway.

So I chose code simplicity over performance here. Do you think we should go for the log method?
(Current implementation has complexity log(length) which is still not bad.)